runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala

```diff
@@ -0,0 +1,102 @@
+package ch.epfl.lamp.grading
+
+import java.io.{ File, PrintWriter }
+import scala.Predef.{ any2stringadd => _, _ }
+import scala.util.Try
+import scala.collection.mutable
+import scala.pickling.directSubclasses
+import scala.pickling.Defaults._
+import scala.pickling.json._
+import Entry._
+
+final case class GradingSummary(score: Int, maxScore: Int, feedback: String)
+object GradingSummary {
+  class StringSplitExtractor(splitter: String) {
+    def unapplySeq(str: String) = Array.unapplySeq(str.split(splitter))
+  }
+  object ToInt {
+    def unapply(s: String) = Try(s.toInt).toOption
+  }
+
+  /** suiteName::suiteWeight */
+  object SuiteId extends StringSplitExtractor("::")
+
+  /** suiteName::testName::testWeight */
+  object TestId extends StringSplitExtractor("::")
+
+  final case class Suite(val name: String, val weight: Int,
+    var complete: Boolean = false,
+    val tests: mutable.Map[String, Test] = mutable.Map.empty)
+
+  final case class Test(val name: String, val weight: Int,
+    var failure: Option[String])
+
+  /** Given test completion state, compute the grade and provide feedback. */
+  def apply(suites: List[Suite]): GradingSummary = {
+    val score = suites.map { _.tests.values.map { t => t.failure.fold(t.weight)(_ => 0) }.sum }.sum
+    val maxScore = suites.map { _.weight }.sum
+    val feedback = {
+      val sb = new StringBuilder
+      sb.append {
+        "Your solution achieved a testing score of %d out of %d.\n\n".format(score, maxScore)
+      }
+      if (score == maxScore)
+        sb.append("Great job!!!\n\n")
+      else {
+        sb.append("""Below you can see a short feedback for every test that failed,
+                    |indicating the reason for the test failure and how many points
+                    |you lost for each individual test.
+                    |
+                    |Tests that were aborted took too long too complete or crashed the
```

Comment:
It refers to "test has been aborted" failure. How would you reformulate the sentence? It's not quite clear to me.